README
This is Perl module Kagura.

INSTALLATION

Kagura installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Kagura

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Kagura documentation is available as in POD. So you can do:

    % perldoc Kagura

to read the documentation online with your favorite pager.

xaicron


Changes
Revision history for Perl extension Kagura

0.01  Sun Jan 23 23:30:05 2011
    - original version


script/kagura-setup.pl
#!perl

use strict;
use warnings;
use Text::MicroTemplate qw(render_mt);
use Getopt::Long qw(GetOptions);
use File::Path qw/mkpath/;
use File::Basename qw/dirname/;
use File::Spec;
use Kagura ();

my $renderer = 'Text::MictoTemplate::File';
my $suffix_map = +{
    'Text::MictoTemplate::File' => 'mt',
    'Text::Xslate'              => 'xt',
};

GetOptions(
    'h|help!'             => sub { usage() },
    'r|renderer=s'        => \$renderer,
    's|template-suffix=s' => \my $suffix,
    'p|plugin=s@'         => \my @plugins,
) or usage();

my $name = shift || usage();
(my $base_dir  = $name) =~ s{::}{-}g;
(my $base_path = $name) =~ s{::}{/}g;

main: {
    die "$base_dir is exists!\n" if -e $base_dir;
    mkdir $base_dir or die "$base_dir: $!";
    chdir $base_dir or die "$base_dir: $!";

    write_all();
    exit;
}

sub write_all {
    my $data = _parse_data_section();

    print "Generating for $name\n";
    for my $path (sort keys %$data) {
        my $dir = dirname($path);
        unless (-e $dir) {
            mkpath($dir) or die "Cannot mkpath '$dir': $!";
        }

        print "writing $path\n";
        my $content = $data->{$path};
#        if ($path =~ /\.(?:ico|jpe?g|png|gif)$/) {
#            $content = MIME::Base64::Perl::decode_base64($content);
#        }
        open my $out, '>', $path or die "Cannot open '$path' for writing: $!";
        print $out $content;
        close $out;
    }
}

sub _parse_data_section {
    my $data_string = render_mt(do { local $/; <DATA> }, +{
        kagura_version => $Kagura::VERSION,
        name           => $name,
        base_dir       => $base_dir,
        base_path      => $base_path,
        renderer       => $renderer,
        suffix         => $suffix || $suffix_map->{$renderer} || 'mt',
        plugins        => [@plugins],
    });

    my ($data, $path);
    for my $line (split /\n/, $data_string) {
        if ($line =~ /^\@\@/) {
            ($path) = $line =~ /^\@\@ (.*)/;
            next;
        }
        next unless $path;
        $data->{$path} .= "$line\n";
    }
    return $data;
}

sub usage {
    print << 'USAGE';
Usage: kagura-setup.pl [options] MyApp

Options:
    h, help         show this message
    r, renderer     set rederer class (e.g. Text::Xslate. default Text::MicroTemplate)
    suffix          set template suffix (e.g. tt. default is renderer default suffix)
    p, plugin       sets using plugin(s)
    json            added render_json() method

USAGE
    exit 1;
}

=pod

=encoding utf-8

=for stopwords

=head1 NAME

kagura-setup.pl - skeleton generator for Kagura

=head1 SYNOPSIS

  $ kagura-setup.pl MyApp

=head1 AUTHOR

xaicron

=head1 COPYRIGHT

Copyright 2011 - xaicron

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

__DATA__
? my ($params) = @_;
? my $kagura_version = $params->{kagura_version} || 0.01;
? my $name           = $params->{name};
? my $base_dir       = $params->{base_dir};
? my $base_path      = $params->{base_path};
? my $renderer       = $params->{renderer};
? my $suffix         = $params->{suffix};
? my $plugins        = $params->{plugins};

@@ <?= $base_dir ?>.psgi
use lib 'lib';
use <?= $name ?>;
use Plack::Builder;

my $app = <?= $name ?>->to_app;
builder {
    enable 'Static',
        path => qr{^/static|^/favicon.ico$},
        root => './htdocs';
    $app;
};

@@ conf/development.pl
+{
};

@@ lib/<?= $base_path ?>.pm
package <?= $name ?>;
use strict;
use warnings;
use <?= $name ?>::Web;
our $VERSION = '0.01';

get '/' => dispatch(Root => 'index');

1;

@@ lib/<?= $base_path ?>/Web.pm
package <?= $name ?>::Web;

use strict;
use warnings;
use parent 'Kagura';

? if ($renderer eq 'Text::Xslate') {
sub init_renderer {
    my ($class) = @_;

    my $config    = $class->config->{template};
    my $path      = $class->home_dir->subdir($config->{path} || 'tmpl');
    my $cache_dir = $path->subdir($config->{cache_dir} || 'cache');

    my $renderer = Tiffany->load('Text::Xslate', {
        syntax    => $config->{syntax} || 'TTerse',
        path      => $path->stringify,
        module    => [ @{ $config->{module} || [] } ],
        cache     => $config->{cache} || 1,
        cache_dir => $cache_dir->stringify,
    });
    $class->renderer($renderer);
}

? }
sub init_prepare {
    my ($class) = @_;
}

sub init_finalize {
    my ($class) = @_;
? for my $plugin (@$plugins) {
    $class->load_plugin('<?= $plugin ?>');
? }
}

1;

@@ lib/<?= $base_path ?>/Web/C/Root.pm
package <?= $name ?>::Web::C::Root;

use strict;
use warnings;

sub index {
    my ($c, $req) = @_;
    $c->render('index.<?= $suffix ?>');
};

1;

@@ lib/<?= $base_path ?>/M.pm
package <?= $name ?>::M;

use strict;
use warnings;

1;

@@ t/00_complie.t
use strict;
use warnings;
use Test::More tests => 1;

BEGIN { use_ok '<?= $name ?>' }

@@ xt/01_podspell.t
use strict;
use warnings;
use Test::More;
use Test::Requires 'Test::Spelling';
use Config;
use File::Spec;
use ExtUtils::MakeMaker;

my %cmd_map = (
    spell  => 'spell',
    aspell => 'aspell list',
    ispell => 'ispell -l',
);

my $spell_cmd;
for my $dir ((split /$Config::Config{path_sep}/, $ENV{PATH}), '.') {
    next if $dir eq '';
    ($spell_cmd) = map { $cmd_map{$_} } grep {
        my $abs = File::Spec->catfile($dir, $_);
        -x $abs or MM->maybe_command($abs);
    } keys %cmd_map;
    last if $spell_cmd;
}
$spell_cmd = $ENV{SPELL_CMD} if $ENV{SPELL_CMD};
plan skip_all => "spell command are not available." unless $spell_cmd;
add_stopwords(map { split /[\s\:\-]/ } <DATA>);
set_spell_cmd($spell_cmd);
$ENV{LANG} = 'C';
all_pod_files_spelling_ok('lib');

__DATA__
<?= $name ?>

@@ xt/02_pod.t
use strict;
use warnings;
use Test::More;

eval "use Test::Pod 1.00";
plan skip_all => "Test::Pod 1.00 required for testing POD" if $@;

all_pod_files_ok();

@@ xt/03_perlcritic
use strict;
use warnings;
use Test::More;

unless ($ENV{TEST_PERLCRITIC}) {
    plan skip_all => "\$ENV{TEST_PERLCRITIC} is not set.";
    exit;
}

eval {
    require Test::Perl::Critic;
    Test::Perl::Critic->import( -profile => 'xt/perlcriticrc');
};
plan skip_all => "Test::Perl::Critic is not installed." if $@;

all_critic_ok('lib');

@@ xt/perlcriticrc
[TestingAndDebugging::ProhibitNoStrict]
allow=refs
[-Subroutines::ProhibitSubroutinePrototypes]

@@ Makefile.PL
use inc::Module::Install;
name '<?= $base_dir ?>';
all_from 'lib/<?= $base_path ?>.pm';

requires 'Kagura', '<?= $kagura_version ?>';
? if ($renderer eq 'Text::Xslate') {
requires 'Text::Xslate', 1.0000;
? }

test_requires 'Test::More', 0.96;
test_requires 'Test::Requires', 0.06;

tests join q{ }, map { sprintf 't%s.t', '/*' x $_ } 1..3;
author_tests 'xt';

WriteAll;

@@ MANIFEST.SKIP
\bRCS\b
\bCVS\b
\.svn/
\.git/
^MANIFEST\.
^Makefile$
~$
\.old$
^blib/
^pm_to_blib
^MakeMaker-\d
\.gz$
\.shipit
\.gitignore
\ppport.h

@@ Changes
Revision history for Perl extension <?= $name ?>

0.01  <?= scalar localtime ?>
    - original version

@@ README
Generated by kagura-setup.pl ($Kagura::VERSION <?= $kagura_version ?>).

You can running this project are:
$ plackup <?= $base_dir ?>.psgi

enjoy!

@@ .shipit
steps = FindVersion, ChangeVersion, CheckChangeLog, DistTest, Commit, Tag, MakeDist, UploadCPAN
git.push_to = origin

@@ .gitignore
cover_db
META.yml
Makefile
blib
inc
pm_to_blib
MANIFEST
Makefile.old
nytprof*
ppport.h
xs/*c
xs/*o
xs/*obj
*.bs
*.def
*.old
dll*
*~

@@ tmpl/index.<?= $suffix ?>
<html>
<head>
    <title>Hello, <?= $name ?></title>
</head>
<body>
    <h1>Hello, <?= $name ?>!!</h1>
    <ul>
        <li>
            Renderer
            <ul>
                <li><?= $renderer ?></li>
            </ul>
        </li>
        <li>
            Template Suffix
            <ul>
                <li><?= $suffix ?></li>
            </ul>
        </li>
? if (@$plugins) {
        <li>
            Plugins
            <ul>
?   for my $plugin (@$plugins) {
                <li><?= $plugin ?></li>
?   }
            </ul>
        </li>
? }
    </ul>
    <p>$Kagura::VERSION: <?= $kagura_version ?>
</body>
</html>



lib/Kagura/Plugin/Logger.pm
package Kagura::Plugin::Logger;

use strict;
use warnings;
use Kagura::Util;

sub init {
    my ($class, $c, $conf) = @_;
    $conf ||= $c->config->{plugin}{Logger} || {};

    my $logger;
    my $args = $conf->{args} || [];
    if (my $logger_class = $conf->{class}) {
        $logger = Plack::Util::load_class($logger_class)->new(@$args);
    }
    else {
        $logger = Kagura::Log->new(@$args);
    }

    $c->mk_classdata('logger');
    $c->logger($logger);
    Kagura::Util::add_method($c, log => sub { $logger });
}

# no index
package
    Kagura::Log;

use parent 'Log::Dispatch';
use Data::Dump ();
use POSIX qw(strftime);

BEGIN {
    for my $level (qw{
        debug info notice warn warning err error crit critical alert emerg
    }) {
       my $sub = sub {
            my $self = shift;
            my ($module, $file, $line);
            my $i = 0;
            while (($module, $file, $line) = caller($i++)) {
                last if $module !~ m{^(?:Kagura::Plugin::Log::Dispatch|Log::Dispatch)};
            }
            $self->log(
                level   => $level,
                message => "@_",
                module  => $module,
                file    => $file,
                line    => $line,
            );
       };

       no strict 'refs';
       *{$level} = $sub;
    }
}

sub new {
    my ($class, %args) = @_;
    unless ($args{outputs}) {
        $args{outputs} = [
            [ 'Screen', min_level => 'debug', stderr => 1, newline => 1 ],
        ];
    }
    $class->SUPER::new(%args);
}

sub dump {
    my ($self, @data) = @_;
    $self->debug(Data::Dump::dump(@data));
}

sub _log_to {
    my ($self, %p) = @_;
    $p{pid} = $$;
    if ($p{name} eq 'Syslog') {
        $p{message} = sprintf( '[%s] [%s] %s (file: %s, line: %d, pid: %d)',
            @p{qw/level module message file line pid/} );
    }
    else {
        $p{datetime} = strftime( '%Y-%m-%d %H:%M:%S', localtime );
        $p{message} = sprintf( '[%s] [%s] [%s] %s (file: %s, line: %d, pid: %d)',
            @p{qw/datetime level module message file line pid/} );
    }
    $self->SUPER::_log_to(%p);
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Kagura::Plugin::Logger - sets logger

=head1 SYNOPSIS

  package MyApp;
  use parent qw(Kagura);
  
  __PACKAGE__->load_plugin('Logger');
  
  # in your config.pl
  plugin => +{
      Logger => +{
          class => 'MyLogger', # default Kagura::Plugin::Log::Dispatch
          args  => [
              outputs => [
                  'Screen' => (
                      min_level => 'debug',
                      stderr    => 1,
                      newline   => 1,
                  ),
              ],
          ],
      },
  },
  
  # in your controller
  $c->log->debug('foo bar');

=head1 DESCRIPTION

Kagura::Plugin::Logger is sets logger class for Kagura.

Any logger class can be set by configuration.

=head1 AUTHOR

xaicron E<lt>xaicron {at} cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2011 - xaicron

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut


lib/Kagura/Plugin/Web/JSON.pm
package Kagura::Plugin::Web::JSON;

use strict;
use warnings;
use Kagura::Util;
use JSON ();

sub init {
    my ($class, $c, $conf) = @_;
    my $json = JSON->new->utf8(1);
    Kagura::Util::add_method($c, render_json => sub {
        my ($self, $data) = @_;
        my $content = $json->encode($data);
        $self->response_class->new(200, [
            'Content-Length' => length($content),
            'Content-Type'   => 'application/json; charset=utf-8',
        ], [$content]);
    });
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Kagura::Plugin::Logger - sets logger

=head1 SYNOPSIS

  package MyApp;
  use parent qw(Kagura);
  
  __PACKAGE__->load_plugin('Web::JSON');
  
  # in your controller
  $c->render_json(+{ foo => 'bar' });

=head1 AUTHOR

xaicron E<lt>xaicron {at} cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2011 - xaicron

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut


lib/Kagura.pm
package Kagura;

use strict;
use warnings;
use Plack::Request;
use Plack::Response;
use Tiffany;
use Encode ();
use Router::Simple::Sinatraish ();
use Class::Data::Inheritable;
use Path::Class qw/file dir/;
use Plack::Util ();
use Object::Container ();
use Class::Accessor::Lite (
    new => 1,
    ro  => [qw/req params/],
    rw  => [qw/stash/],
);

our $VERSION = '0.01';

sub import {
    my ($pkg) = @_;
    my $class = caller(0);

    Router::Simple::Sinatraish->export_to_level(1);

    no strict 'refs';
    *{$class."::to_app"} = \&to_app;
    *{$class."::_class"} = sub { $pkg };
    *{$class."::dispatch"} = sub {
        use strict 'refs';
        dispatch($class, @_);
    };
    *{$class."::init"} = sub {
        use  strict 'refs';
        $pkg->init;
    };
}

sub contenxt { die "no context is awaked" }

sub mk_classdata {
    my $class = shift;
    Class::Data::Inheritable::mk_classdata($class, @_);
}

sub init {
    my ($class) = @_;

    $class->init_prepare();

    $class->mk_classdata('config');
    $class->mk_classdata('home_dir');
    $class->mk_classdata('renderer');
    $class->mk_classdata('container', 'Object::Container');
    $class->mk_classdata('response_class', 'Plack::Response');
    $class->mk_classdata('request_class', 'Plack::Request');
    $class->mk_classdata('_loaded_plugin', +{});

    $class->init_home_dir();
    $class->init_config();
    $class->init_renderer();
    $class->init_container();
    $class->init_plugins();

    $class->init_finalize();
}

# you can override this method
sub init_prepare  {}
sub init_finalize {}

sub init_home_dir {
    my ($class) = @_;
    my $home_dir = dir($ENV{KAGURA_HOME} || '.');
    $class->home_dir($home_dir);
}

sub init_config {
    my ($class) = @_;
    my $env    = $ENV{PLACK_ENV} || 'development';
    my $fname  = $class->home_dir->file('conf', "$env.pl")->stringify;
    my $config = do $fname or die "cannot load configuration file: $fname";
    $class->config($config);
}

sub init_renderer {
    my ($class) = @_;

    my $config = $class->config->{template};
    my $path   = do {
        my $path = $config->{path} || ['tmpl'];
        $path = [ $path ] unless ref $path eq 'ARRAY';
        $path;
    };

    my $renderer = Tiffany->load('Text::MicroTemplate::File', {
        include_path => [ map { $class->home_dir->subdir($_)->stringify } @$path ],
        use_cache    => $config->{use_cache},
    });

    $class->renderer($renderer);
}

sub init_container {
    my ($class) = @_;
    return unless $class->config->{container};

    my $container = $class->config->{container};
    for my $name (keys %$container) {
        $class->container->register({
            class       => $name,
            initializer => $container->{$name}{init},
            args        => $container->{$name}{args},
            preload     => $container->{$name}{preload},
        });
    }
}

sub init_plugins {
    my ($class) = @_;
    my $config = $class->config->{plugin} || {};
    $class->load_plugins(%$config);
}

sub load_plugins {
    my ($class, %args)  = @_;
    for my $module (keys %args) {
        my $conf = $args{$module};
        $class->load_plugin($module, $conf);
    }
}

sub load_plugin {
    my ($class, $module, $conf) = @_;
    $module = Plack::Util::load_class($module, __PACKAGE__.'::Plugin');
    return if $class->_loaded_plugin->{$module};
    $module->init($class, $conf);
    $class->_loaded_plugin->{$module}++;
}

sub render {
    my $self = shift;
    my $content = Encode::encode_utf8($self->renderer->render(@_));
    $self->response_class->new(200, [
        'Content-Length' => length($content),
        'Content-Type'   => 'text/html; charset=utf-8',
    ], [$content]);
}

sub return_404 {
    return $_[0]->response_class->new(404, [], ['404 Not Found']);
}

sub return_403 {
    return $_[0]->response_class->new(403, [], ['403 Forbidden']);
}

sub show_error {
    my ($self, $msg) = @_;
    return $self->response_class->new(500,
        ['Content-Type' => 'text/html; charset=utf-8'],
        [$msg || '500 Internal Server Error'],
    );
}

sub model {
    my ($self, $model) = @_;
    my $model_class = "$self->{class}::M::$model";
    $self->container->get($model_class);
}

sub dispatch {
    my ($class, $route, $method) = @_;
    my $module = Plack::Util::load_class($route, "$class\::Web::C");
    return \&{"$module\::$method"};
}

sub to_app {
    my ($class) = shift;

    my $pkg = $class->_class;
    $pkg->init;

    sub {
        my $env = shift;
        if (my ($match, $route) = $class->router->routematch($env)) {
            delete $match->{code} if ref $match->{code} eq 'CODE';

            my $req = $pkg->request_class->new($env);
            my $c = $pkg->new(
                req    => $req,
                class  => $class,
                stash  => {},
                params => +{ %$match },
            );
            no strict 'refs';
            local *{"$pkg\::context"} = sub { $c };
            use strict 'refs';

            my $res = $route->{dest}{code}->($c, $req);
            $res = $c->show_error() unless ref $res;
            return $res->finalize;
        }
        else {
            return $pkg->return_404();
        }
    };
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Kagura - simple web application framework

=head1 SYNOPSIS

Generating new project:

  $ kagura-setup.pl MyApp

run app.psgi

  $ cd MyApp
  $ plackup MyApp.psgi

=head1 DESCRIPTION

Kagura is easy, simple, lightweight web application framework.

=head1 EXPORT FUNCTIONS

=over 4

=item B<< get($path, $code) >>

=item B<< post($path, $code) >>

=item B<< any([$method, ] $path, $code) >>

This functions from L<< Router::Simple::Sinatraish >>.

C<< $code >> must be returned Plack::Response-like object.

  pacakge MyApp;
  use MyApp::Web;
  
  get '/' => sub {
      my ($c, $req) = @_;
      ...
      my $res = $c->response_class->new(200, [], ['Hello Kagura!!']);
      return $res;
  }
  post '/' => sub {
      my ($c, $req) = @_;
      ...
      return $res;
  }
  any '/' => sub {
      my ($c, $req) = @_;
      ...
      return $res;
  }
  any [qw/GET DELETE/], '/any' => sub {
      my ($c, $req) = @_;
      ...
  }
  
  # $c   is Kagura-like object
  # $req is Plack::Requst-like object (eq $c->req)
  # $res is Plack::Response-like object

=item B<< dispatch($contoroller, $method_name) >>

Sets dispatch rule.

into MyApp.pm

  pacakge MyApp;
  use MyApp::Web;
  
  # call MyApp::Web::C::Root::index
  get '/' => dispatch('Root', 'index');

into MyApp/Web/C/Root.pm

  package MyApp::Web::C::Root;
  
  sub index {
      my ($c, $req) = @_;
      ...
  }

=back

=head1 METHOD

=over 4

=item B<< req() >>

Get request object.

  get '/' => sub {
     my ($c, $req) = @_;
     $c->req; # eq $req
     ...
  }

=item B<< params() >>

Route parameters.

  get '/{user}' => sub {
     my ($c, $req) = @_;
     my $params = $c->params;
     ...
  }
  
  # GET http://localhost/foo
  # $params->{user} eq 'foo'

=item B<< render(@args) >>

Rendering template. Returned C<< response_class >> object.

  get '/' => sub {
     my ($c, $req) = @_;
     ...
     my $res = $c->render('index.mt');
     return $res;
  }

=item B<< model($name) >>

Get C<< MyApp::M::$name >> object.

  pacakge MyApp::M::User;
  sub find {
      my ($self, $user_id) = @_;
      ...
      return $user_name;
  }
  
  package MyApp;
  use parent 'MyApp::Web';
  get '/{user_id}' => sub {
      my ($c, $req) = @_;
      my $user_name = $c->model('User')->find($c->params->{user_id});
      ...
  }

=item B<< return_404() >>

Returned status 404.

  get '/404' => sub {
     my ($c, $req) = @_;
     return $c->return_404();
  }

=item B<< return_403() >>

Returned status 403.

  get '/403' => sub {
     my ($c, $req) = @_;
     return $c->return_403();
  }

=item B<< show_error() >>

Returned status 500.

  get '/error' => sub {
     my ($c, $req) = @_;
     return $c->show_error('oops!!');
  }

=item B<< stash([$hashref]) >>

You can use freely.

=back

=head1 CLASS METHOD

=over 4

=item B<< to_app() >>

into app.psgi

  use MyApp;
  MyApp->to_app;

=item B<< init() >>

Initialize application. This method in to_app() called.

=item B<< init_prepare() >>

Call prepare init() method. This method is in inii() called.

You can override this method.

  pacakge MyApp::Web;
  use parent 'Kagura';
  
  sub init_prepare {
      my ($class) = @_;
      ...
  }

=item B<< init_finalize() >>

Call finalize init() method. This method is in inii() called.

You can override this method.

  pacakge MyApp::Web;
  use parent 'Kagura';
  
  sub init_finalize {
      my ($class) = @_;
      ...
  }

=item B<< load_plugin($class [, $config]) >>

Load plugin class. this method must be calling in init_finalzie() or after init().

  package MyApp::Web;
  use parent 'Kagura'
  
  sub init_finalize {
      my ($class) = @_;
      $class->load_plugin('Web::JSON');
      $class->load_plugin('+MyApp::Plugin::Foo', +{ bar => 'baz' });
  }

=item B<< load_plugins(%args) >>

Load plugins.

  package MyApp::Web;
  use parent 'Kagura';
  
  sub init_finalize {
      my ($class) = @_;
      $class->load_plugins(
        'Web::JSON'           => {},
        '+MyApp::Plugin::Foo' => +{ bar => 'baz' }
      );
  }

or configuration:

  # conf/developlemt.pl
  +{
      plugin => +{
          'Web::JSON'           => +{},
          '+MyApp::Plugin::Foo' => +{ bar => 'baz' },
      },
  }

=back

=head1 DEFAULT CLASS ACCESSOR METHOD

These methods you can calling the after init().

=over 4

=item B<< config([$hash_ref]) >>

Loaded configuration from C<< conf/$ENV{PLACK_ENV}.pl >>

=item B<< home_dir([$home_dir]) >>

Sets value must be L<< Path::Class >>-like object.

=item B<< renderer([$renderer]) >>

Sets value must be L<< Tiffany >>::* object.

=item B<< container([$scalar] >>

Default 'Object::Container'

=item B<< request_class([$scalar]) >>

Default 'Plack::Request'

=item B<< response_class([$scalar]) >>

Default 'Plack::Response'

=back

=head1 AUTHOR

xaicron E<lt>xaicron {at} cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2011 - xaicron

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
